sips/sip-tbd/sip-TBD-token-metadata-update-notifications.md

```diff
@@ -0,0 +1,234 @@
+used through `print`. Existing metadata indexers would receive this event through the [Stacks node
+event-emitter interface](https://github.com/stacks-network/stacks-blockchain/blob/master/docs/event-dispatcher.md#post-new_block),
+parse its contents, and refresh any contracts that were updated.
+
+# Specification
+
+Notification messages for each token class are specified below. Token metadata update notifications
+must be made via a contract call transaction to the [deployed reference
+contract](https://explorer.stacks.co/txid/0xe92af2ea5c11e2e6fde4d31fd394de888070efff23bffad04465c549543abfa2?chain=mainnet)
+or from a call to `print` within any other contract, including the token contract itself.
+
+The message structure was designed to be reusable by other SIPs who wish to establish new
+notification standards in the future (i.e. by varying the `notification` and `payload` key values).
+
+## Fungible Tokens
+
+When a contract needs to notify the network that metadata has changed for a **Fungible Token**, it
+shall call `print` with a tuple with the following structure:
+
+```clarity
+{ notification: "token-metadata-update", payload: { token-class: "ft", contract-id: <token contract id> }}
+```
+
+| Key                   | Value                                                                  |
+|-----------------------|------------------------------------------------------------------------|
+| `notification`        | The string `"token-metadata-update"`                                   |
+| `payload.token-class` | The string `"ft"`                                                      |
+| `payload.contract-id` | The contract id (principal) of the contract that declared the token    |
+
+## Non-Fungible Tokens
+
+When a contract needs to notify the network that metadata has changed for a **Non-Fungible Token**,
+it shall call `print` with a tuple with the following structure:
+
+```clarity
+{ notification: "token-metadata-update", payload: { token-class: "nft", token-ids: (list u100, u101), contract-id: <token contract id> }}
+```
+
+| Key                   | Value                                                                  |
+|-----------------------|------------------------------------------------------------------------|
+| `notification`        | The string `"token-metadata-update"`                                   |
+| `payload.token-class` | The string `"nft"`                                                     |
+| `payload.token-ids`   | A list with the uint token ids that need to be refreshed               |
+| `payload.contract-id` | The contract id (principal) of the contract that declared the tokens   |
+
+# Backwards compatibility
+
+Developers who need to update metadata for contracts that were deployed before this SIP is activated
+could deploy a new one with a function that broadcasts this message or use the deployed contract
+described in [Reference Implementations](#reference-implementations).
```

Comment:
If I'm reading this right, you seem to be suggesting that the contract you provide below can be used to notify indexers about changes to other contracts (e.g. legacy contracts that don't conform to this SIP).  If so, then how are the following handled?

* Who's allowed to call this contract?  Is it anyone?  The code permits anyone; is that intentional and/or desired? 
* Is there a DDoS risk to indexers and to metadata servers if someone floods the blockchains with lots of calls to this contract?
* For the NFT case, is there a risk to indexers of receiving gratuitous updates -- e.g. for NFTs that don't need to be updated?

Reply:
> the contract you provide below can be used to notify indexers about changes to other contracts

Correct

> Who's allowed to call this contract? Is it anyone? The code permits anyone; is that intentional and/or desired?

Yes, anyone can call this contract on behalf of other contracts. We made this intentional decision to allow backwards compatibility i.e. refreshing legacy contract tokens (more on possible abuse below).

> Is there a DDoS risk to indexers and to metadata servers if someone floods the blockchains with lots of calls to this contract?

In theory yes, there could be. If someone emits hundreds of notifications for Megapont, for example (thousands of tokens), then metadata indexers would need to make a whole lot of read-only contract calls potentially DDoS-ing some nodes. However, this would be mitigated by transaction fees that would need to be paid by attackers, and IMO would still be preferable to the current polling configuration that most indexers have.

> For the NFT case, is there a risk to indexers of receiving gratuitous updates -- e.g. for NFTs that don't need to be updated?

Not really, it would be an idempotent operation as they would get the same results they had before.